docs/zkapps/advanced-snarkyjs.mdx

```diff
@@ -291,3 +291,170 @@ We imagine that one of the most important use cases for actions is to enable con
+we now have access to more data off-chain!
+
+Let's imagine a zkApp that manages a game with a leaderboard. The zkApp has a method to update the score of a player, if the player guesses a number correctly.
+Once a player reaches a threshold score, they can invoke another method to get a reward. Because we want many players to participate in our game,
+we are drastically limited by how much data we can store on-chain - and if we want to have at least 8 or more participants, we will run out of on-chain space very fast!
+
+A possible solution to that problem would be to utilize the power of Merkle trees and store the public keys of each player and their corresponding scores off-chain and only reference it in our smart contract.
+
+Lets look at our data structure first. Imagine wanting to map a player's id to score points, so this:
+
+```
+0: 5 points
+1: 3 points
+2: 0 points
+3: 8 points
+... : ...
+7: 2 points
+```
+
+#### Implementing the smart contract
+
+Now it's time to look at what our leaderboard zkApp might look like. We want to have on-chain state that points to the off-chain Merkle tree - we will call this variable the `root`.
+
+<Alert kind="note">
+
+Sometimes the variable `root` is also called commitment, because we commit to something!
+
+</Alert>
+
+Additionally, we want to store a variable `z` which is the hash of the value a player has to guess: `H(guess) = z`
+
+The first method allows a player to make a guess, and if the guess is correct, the player will gain one point.
+The method takes the players guess and hashes it, it then checks if the hash H(guess) equals the on-chain state z and if that's the case,
+then the player gains one point on the score board.
+
+We will also need a second method that takes care of the reward. It checks if the players score is over a threshold and pays out a reward if that's the case.
+This method will also have to verify our Merkle witness and check it if it matches the on-chain stored Merkle root!
+
+<Alert kind="note">
+
+A fully working example, including all the necessary boilerplate code, can be found in the SnarkyJS repository in the examples folder [here](https://github.com/o1-labs/snarkyjs/tree/main/src/examples/zkapps/merkle_tree).
+
+</Alert>
+
+```ts
+class Leaderboard extends SmartContract {
+  // the root it the root hash of our off-chain Merkle tree
+  @state(Field) root = State<Field>();
+
+  // z is the hashed number we want to guess!
+  @state(Field) z = State<Field>();
+
+  deploy(args: DeployArgs) {
+    super.deploy(args);
+
+    // this is our hash we want to guess! its the hash of the preimage "22", but keep it a secret!
+    this.z.set(
+      Field(
+        "17057234437185175411792943285768571642343179330449434169483610110583519635705"
+      )
+    );
+  }
+
+  @method
+  guessPreimage(guess: Field, account: Account, path: MerkleWitness) {
```

Comment:
Can you use `const` where possible please? I know Gregor likes to use `let` everywhere, but I'd prefer if we get away from doing this b/c doing so looks a bit sloppy.